scripts/gen_shadow_yaml.sh

```diff
@@ -0,0 +1,305 @@
+  printf "  graph:\n"
+  printf "    type: gml\n"
+  printf "    inline: |\n"
+  printf "      graph [\n"
+  printf "        directed 0\n"
+
+  # Print node entries for each host
+  for ((i=0; i<NODE_COUNT; i++)); do
+    printf "        node [\n"
+    printf "          id %d\n" "$i"
+    printf "          host_bandwidth_up \"%s\"\n" "$BANDWIDTH_HOST"
+    printf "          host_bandwidth_down \"%s\"\n" "$BANDWIDTH_HOST"
+    printf "        ]\n"
+  done
+
+  # Central switch node (id = NODE_COUNT)
+  central_id=$NODE_COUNT
+  printf "        node [\n"
+  printf "          id %d\n" "$central_id"
+  printf "          host_bandwidth_up \"%s\"\n" "$BANDWIDTH_SWITCH"
+  printf "          host_bandwidth_down \"%s\"\n" "$BANDWIDTH_SWITCH"
+  printf "        ]\n"
+
+  # Self-loop edges for hosts and switch
+  for ((i=0; i<=NODE_COUNT; i++)); do
+    printf "        edge [\n"
+    printf "          source %d\n" "$i"
+    printf "          target %d\n" "$i"
+    printf "          latency \"%s\"\n" "$LINK_LATENCY"
+    printf "          packet_loss %s\n" "$PACKET_LOSS"
+    printf "        ]\n"
+  done
+
+  # Edges from each host to central switch
+  for ((i=0; i<NODE_COUNT; i++)); do
+    printf "        edge [\n"
+    printf "          source %d\n" "$i"
+    printf "          target %d\n" "$central_id"
+    printf "          latency \"%s\"\n" "$LINK_LATENCY"
+    printf "          packet_loss %s\n" "$PACKET_LOSS"
+    printf "        ]\n"
+  done
+
+  printf "      ]\n"
+
+  printf "hosts:\n"
+
+  for ((i=0; i<NODE_COUNT; i++)); do
+    key_file="${NODE_KEY_FILES[$i]}"
+    node_name="node$i"
+    # Prefer the IP from validators.yaml if present; otherwise fall back to generated 10.0.0.<base+idx>
+    if [[ -n "${VALIDATOR_IPS[$i]}" ]]; then
+      ip="${VALIDATOR_IPS[$i]}"
+    else
+      ip_last=$((IP_BASE_LAST_OCTET + i))
+      ip="10.0.0.$ip_last"
+    fi
+
+    # Prefer the quic port from validators.yaml if present; otherwise use UDP_BASE + index
+    if [[ -n "${VALIDATOR_QUICS[$i]}" ]]; then
+      udp_port="${VALIDATOR_QUICS[$i]}"
+    else
+      udp_port=$((UDP_BASE + i))
+    fi
+
+    prom_port=$((PROM_BASE + i))
+
+    # Build args string
+    args_str=(
+      "--modules-dir" "$MODULES_DIR_ABS"
+      "--bootnodes" "$NODES_YAML"
+      "--genesis" "$CONFIG_YAML"
+      "--validator-registry-path" "$VALIDATORS_YAML"
+      "--node-id" "node_${i}"
+      "--node-key" "$key_file"
+      "--listen-addr" "/ip4/0.0.0.0/udp/${udp_port}/quic-v1"
+      "--prometheus-port" "$prom_port"
+    )
+    # Join args preserving spaces
+    IFS=' ' read -r -a _dummy <<< "" # reset
+    joined="${args_str[*]}"
+
+    printf "  %s:\n" "$node_name"
+    printf "    network_node_id: %d\n" "$i"
+    printf "    ip_addr: %s\n" "$ip"
+    printf "    processes:\n"
+    printf "      - path: %s\n" "$QLEAN_PATH_ABS"
+    printf "        args: \"%s\"\n" "$(yaml_escape "$joined")"
```

Comment:
Better to build list of args: it mor resistant to spaces in paths and special characters